base/cvd/cuttlefish/host/libs/web/credential_source.cc

```diff
@@ -108,28 +189,27 @@ Result<std::unique_ptr<CredentialSource>> GetCredentialSourceLegacy(
   return result;
 }
 
-}  // namespace
-
 GceMetadataCredentialSource::GceMetadataCredentialSource(
     HttpClient& http_client)
-    : http_client(http_client) {
-  latest_credential = "";
-  expiration = std::chrono::steady_clock::now();
+    : http_client_(http_client) {
+  latest_credential_ = "";
+  expiration_ = std::chrono::steady_clock::now();
 }
 
 Result<std::string> GceMetadataCredentialSource::Credential() {
-  if (expiration - std::chrono::steady_clock::now() < kRefreshWindow) {
+  std::lock_guard lock(latest_credential_mutex_);
+  if (expiration_ - std::chrono::steady_clock::now() < kRefreshWindow) {
     CF_EXPECT(RefreshCredential());
   }
-  return latest_credential;
+  return latest_credential_;
```

Comment:
nit: This code is repeated among all 3 implementations. `Credential()`, `latest_credential_` and `latest_credential_mutex_` could be in the base class, where `RefreshCredential()` is declared as pure virtual and implemented by the subclasses.

Reply:
Technically there are 4 implementations, `FixedCredentialSource` doesn't have this logic. You are correct that it's repeated and there could be some code reuse.

Reply:
Added an intermediate `class RefreshingCredentialSource` that holds the state and refresh logic.